WordPress/gutenberg
# Site Editor: Persistent List View

## Description

Fixes #27395

Convert the List View dropdown into a persistent sidebar occupying the same interface skeleton slot as the Inserter (`secondarySidebar`).

To achieve this, a number of widespread changes have been made:

- Create a new `components/secondary-sidebar` folder containing both the Inserter and the List View file, making the Site Editor's `editor.js` easier to read.
- Create a new `listViewPanel` reducer (and relative action and selector) in the same vein as those for the inserter panel and navigation sidebar. These three reducers are interconnected: only one can be open at any time, so the action that opens one will automatically close the others.
- Create a new `listView` icon containing the List View icon, which at this point would have been hardcoded and repeated in 3 different files. Also replace those hardcoded icons with the new one.
- Replace the use of `BlockNavigationDropdown` in the Site Editor header with a custom implementation that would fill the `secondarySidebar` with the `BlockNavigationTree` component, displaying the entire blocks hierarchy of the content, as opposed to the dropdown, which only shows the currently selected nested hierarchy (e.g. if the site title block is selected, the dropdown would only show up to the header template part).
- Clicking on an item in the List View sidebar will select the block in the canvas (This is a pre-existing functionality of `BlockNavigationTree` itself).
- Reuse the same toggle keyboard shortcut for the post editor List View dropdown: <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> (letter o, not number 0; see also the [issue](https://github.com/WordPress/gutenberg/issues/27395#issuecomment-738732612)).
- Add a new `highlightBlocksOnHover` prop to the `BlockNavigation` component to support highlighting the corresponding block in the canvas on hover and focus.

### TODO

- Accessibility!
- Highlight the list view item when hovering the block in the canvas (candidate for a follow up: https://github.com/WordPress/gutenberg/issues/28699).
- Figure out a sane keyboard navigation (more on this below).

### Stuff to discuss

The keyboard navigation.
This is complicated, and any changes will affect the original `BlockNavigation` component itself.
What makes things awkward is that selecting a block in the `BlockNavigation` moves the focus to the canvas, and selecting blocks in the canvas will change the `BlockNavigation` selected item.
A common scenario would be:
1. Open the sidebar.
2. Move with the up/down keys: this only highlights the items with a border, not select them.
3. Hit enter: this will select the item (it becomes black), and move the focus to the block in the canvas (without any obvious visual clues).
4. Move with up/down: this will actually move between blocks, which in turns moves the selected item in the sidebar.
5. The user (me in this case) is now lost. The only way to return the focus to the sidebar is toggle it off and on again.

In other words: once the user selects an item in the sidebar, the vertical traversal changes.
Whereas one would expect that traversing the block list is just a matter of going up/down, traversing the canvas is a different matter.
Horizontally aligned blocks might want an horizontal movement instead, practically blocking the traversal.

## How has this been tested?

- Install and activate a FSE-enabled theme such as TT1-Blocks.
- Open the Site Editor.
- Click on the "List View" icon, or use the <kbd>Ctrl</kbd>+<kbd>Alt</kbd>+<kbd>O</kbd> shortcut.
- Make sure the List View sidebar pops up.
- Play around with it to see if it works (or not) as expected.
- Make sure only one of these sidebar can be open at any time: list view, inserter, navigation.

## Screenshots

(Ignore the unstyled canvas, it's an unrelated issue)

<img width="864" alt="Screenshot 2021-02-02 at 12 37 30" src="https://user-images.githubusercontent.com/2070010/106601379-797f2a00-6553-11eb-9e5e-9d500fe3673b.png">

## Types of changes
New feature (non-breaking change which adds functionality)

## Checklist:
